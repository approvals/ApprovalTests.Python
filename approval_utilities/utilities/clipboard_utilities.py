def copy_to_clipboard(text: str) -> None:
    """
    This acts as a wrapper to defer the importing of pyperclip util actual usage
    """
    import pyperclip

    pyperclip.copy(text)
