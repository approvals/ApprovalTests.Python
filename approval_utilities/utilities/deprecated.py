import functools
import warnings
from typing import Callable


def deprecated(reason: str) -> Callable:
    def decorator(func: Callable) -> Callable:
        """This is a decorator which can be used to mark functions
        as deprecated. It will result in a warning being emitted
        when the function is used."""

        @functools.wraps(func)
        def new_func(*args, **kwargs):
            warnings.simplefilter("always", DeprecationWarning)  # turn off filter
            warnings.warn(
                f"Call to deprecated function {func.__name__}. {reason}",
                category=DeprecationWarning,
                stacklevel=2,
            )
            warnings.simplefilter("default", DeprecationWarning)  # reset filter
            return func(*args, **kwargs)

        return new_func

    return decorator
