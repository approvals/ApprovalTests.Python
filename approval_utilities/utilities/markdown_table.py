from typing import TYPE_CHECKING, Any, Callable, Iterable

from typing_extensions import override

from approval_utilities.approvaltests.core.verifiable import Verifiable
from approval_utilities.approvaltests.core.verify_parameters import VerifyParameters

if TYPE_CHECKING:
    from approvaltests.core.options import Options


class MarkdownTable(Verifiable):
    def __init__(self) -> None:
        self.markdown: str = ""

    @override
    def get_verify_parameters(self, options: "Options") -> VerifyParameters:
        return VerifyParameters(options.for_file.with_extension(".md"))

    @staticmethod
    def with_headers(*column_names: str) -> "MarkdownTable":
        table = MarkdownTable()
        table.add_rows(*column_names)
        dividers = map(lambda _: "---", column_names)
        table.add_rows(*dividers)
        return table

    def add_rows(self, *column_names: str) -> "MarkdownTable":
        self.markdown += MarkdownTable.print_row(*column_names)
        return self

    @override
    def __str__(self) -> str:
        return self.markdown

    @staticmethod
    def print_row(*column_names: Any) -> str:
        row = "| " + " | ".join(map(str, column_names)) + " |"
        return row + "\n"

    def add_rows_for_inputs(
        self, inputs: Iterable[Any], *input_transformers: Callable[[Any], Any]
    ) -> "MarkdownTable":
        def transform_resolver_for_input(
            input_value: Any,
        ) -> Callable[[Callable[[Any], Any]], Any]:
            return lambda transform: transform(input_value)

        for row_input in inputs:
            row = [row_input]
            row += map(transform_resolver_for_input(row_input), input_transformers)
            self.add_rows(*row)
        return self
