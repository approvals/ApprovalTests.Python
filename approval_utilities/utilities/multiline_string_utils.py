import textwrap
from typing import Optional


def remove_indentation_from(text: Optional[str]) -> str:
    if not text:
        return ""
    cleaned = textwrap.dedent(text + "|")
    if cleaned.startswith("\n"):
        cleaned = cleaned[1:]
    return cleaned[:-1]
