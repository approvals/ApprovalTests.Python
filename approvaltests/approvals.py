from threading import local
import xml.dom.minidom

from approvaltests import to_json
from approvaltests.approval_exception import ApprovalException
from approvaltests.core.namer import Namer
from approvaltests.core.scenario_namer import ScenarioNamer
from approvaltests.core.pytest_namer import PytestNamer
from approvaltests.file_approver import FileApprover
from approvaltests.list_utils import format_list
from approvaltests.reporters.diff_reporter import DiffReporter
from approvaltests.string_writer import StringWriter

DEFAULT_REPORTER = local()


def set_default_reporter(reporter):
    global DEFAULT_REPORTER
    DEFAULT_REPORTER.v = reporter
    

def get_default_reporter():
    if not hasattr(DEFAULT_REPORTER, "v") or DEFAULT_REPORTER.v is None:
        return DiffReporter()
    return DEFAULT_REPORTER.v


def get_reporter(reporter):
    return reporter or get_default_reporter()


def get_default_namer():
    return Namer()


def verify(data, reporter=None, namer=None):
    reporter_to_use = reporter or get_default_reporter()
    namer_to_use = namer or get_default_namer()
    verify_with_namer(data, namer_to_use, reporter_to_use)


def verify_with_namer(data, namer, reporter):
    reporter = get_reporter(reporter)
    approver = FileApprover()
    writer = StringWriter(data)
    error = approver.verify(namer, writer, reporter)
    if error is not None:
        raise ApprovalException(error)


def verify_as_json(object, reporter=None):
    n_ = to_json(object) + "\n"
    verify(n_, reporter)


def verify_xml(xml_string, reporter=None):
    try:
        dom = xml.dom.minidom.parseString(xml_string)
        pretty_xml = dom.toprettyxml()
    except Exception:
        pretty_xml = xml_string
    verify_with_namer(pretty_xml, Namer(extension='.xml'), reporter)


def verify_file(file_name, reporter=None):
    with open(file_name, 'r') as f:
        file_contents = f.read()
    verify(file_contents, reporter)


def verify_all(header, alist, formatter=None, reporter=None):
    text = format_list(alist, formatter, header)
    verify(text, reporter)


def get_scenario_namer(scenario_name):
    return ScenarioNamer(get_default_namer(), scenario_name)

def get_pytest_namer(pytest_request):
    return PytestNamer(get_default_namer(), pytest_request)