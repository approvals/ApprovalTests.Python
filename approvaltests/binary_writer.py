from typing import ByteString

from typing_extensions import override

from approval_utilities.utils import create_directory_if_needed
from approvaltests.core.writer import Writer


class BinaryWriter(Writer):
    contents: ByteString

    def __init__(
        self,
        contents: ByteString,
        extension: str,
    ) -> None:
        self.contents = contents
        self.extension_with_dot = extension

    @override
    def write_received_file(self, received_file: str) -> str:
        create_directory_if_needed(received_file)
        with open(received_file, mode="wb") as file:
            file.write(self.contents)

        return received_file
