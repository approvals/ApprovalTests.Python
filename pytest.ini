[pytest]
testpaths = tests
addopts = --showlocals

asyncio_default_fixture_loop_scope = function
