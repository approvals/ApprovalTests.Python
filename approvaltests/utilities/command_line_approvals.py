import os
import subprocess
from typing import Any, Dict, Optional, Sequence

from approvaltests import Options, verify


def verify_command_line_with_inputs(
    command: str,
    *,
    inputs: Optional[Sequence[Any]] = None,
    options: Optional[Options] = None,
) -> None:
    input_string = "\n".join(map(lambda a: f"{a}", inputs))
    verify_command_line(command, input_string=input_string, options=options)


def verify_command_line(
    command_line: str,
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    input_string: Optional[str] = None,
    options: Optional[Options] = None,
    current_working_directory: str = ".",
    additional_environment_variables: Optional[Dict[str, str]] = None,
) -> None:
    # Set up environment with UTF-8 support for Windows
    my_env = {**os.environ}

    # Ensure UTF-8 encoding on Windows for emoji support
    if os.name == "nt":  # Windows
        my_env.update({"PYTHONIOENCODING": "utf-8", "PYTHONUTF8": "1"})

    if additional_environment_variables:
        my_env.update(additional_environment_variables)
    try:
        result = subprocess.run(
            command_line,
            shell=True,
            text=True,
            encoding="utf-8",
            input=input_string,
            cwd=current_working_directory,
            env=my_env,
            capture_output=True,
        )
        output = result.stdout
        if result.stderr:
            output += "\n--- STDERR ---\n" + result.stderr
    except UnicodeDecodeError:
        # Fallback to system default encoding if UTF-8 fails
        result = subprocess.run(
            command_line,
            shell=True,
            text=True,
            input=input_string,
            cwd=current_working_directory,
            env=my_env,
            capture_output=True,
        )
        output = result.stdout
        if result.stderr:
            output += "\n--- STDERR ---\n" + result.stderr
    verify(
        output,
        options=options,
    )
