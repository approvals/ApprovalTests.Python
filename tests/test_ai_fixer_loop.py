from pathlib import Path
import re
import sys
import pytest

from approvaltests import Options
from approvaltests.utilities.command_line_approvals import verify_command_line

_SCRIPT_DIR = Path(__file__).parent
_REPO_ROOT = _SCRIPT_DIR.parent


# @pytest.mark.skip
def test_ai_fixer_loop() -> None:
    python = "python"
    cwd = _REPO_ROOT / "internal_documentation" / "scripts"
    verify_command_line(
        python + " ai_fixer_loop.py --find test_find --fix test_fix --tcr test_tcr",
        current_working_directory=str(cwd),
        options=Options().with_scrubber(
            lambda text: re.sub(r" \[[\d\.]+s\]", " [time]", text)
        ),
    )
