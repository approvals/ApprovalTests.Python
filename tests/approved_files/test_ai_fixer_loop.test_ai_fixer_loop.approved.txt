Running initial check...
CHECKMARK tcr: committed [time]
CHECKMARK build working
-----------------------------
Run #1
CHECKMARK found problems [time]
CHECKMARK fix problem [time]
CHECKMARK tcr: committed [time]
-----------------------------
Run #2
CHECKMARK found problems [time]
CHECKMARK fix problem [time]
X tcr: reverted [time]
-----------------------------
Run #3
CHECKMARK found problems [time]
CHECKMARK fix problem [time]
CHECKMARK tcr: committed [time]
-----------------------------
Run #4
X found problems [time]
no more problems found.
